californium-core/src/test/java/org/eclipse/californium/core/test/lockstep/LockstepEndpoint.java

```diff
@@ -572,17 +669,99 @@ public EmptyMessageExpectation(Type type, int mid) {
 		}
 
 		@Override
+		public void go(Message msg) throws Exception {
+			if (CoAP.isEmptyMessage(msg.getRawCode())) {
+				check(msg);
+			} else {
+				Assert.fail("Expected empty message for " + this + ", but received " + msg);
+			}
+		}
+
+		@Override
+		public void add(MultiMessageExpectation multi) {
+			multi.add(this);
+		}
+	}
+
+	public class MultiMessageExpectation implements Action {
+
+		private int counter;
+		private EmptyMessageExpectation emptyExpectation;
```

Comment:
My first impression was, that calling "MessageExpectation,check(Message)" is not free of side effects.
There is logging output and the "store" methods. So just iterate over all possible "Expectations" and call "check" to find one, which succeeds, seem for me to introduce some more noise.
Therefore I chose to use the tpyed approach to select the message check.
This will only work, as long as only ACKs and Responses are assumed to be send "at once" without further "step" interaction.